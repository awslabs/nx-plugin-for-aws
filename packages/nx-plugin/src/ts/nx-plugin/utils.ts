/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  addDependenciesToPackageJson,
  joinPathFragments,
  readJson,
  Tree,
  updateJson,
  writeJson,
} from '@nx/devkit';
import { readProjectConfigurationUnqualified } from '../../utils/nx';
import { withVersions } from '../../utils/versions';
import PackageJson from '../../../package.json';

/**
 * Configures a TypeScript project as an Nx Plugin
 */
export const configureTsProjectAsNxPlugin = (
  tree: Tree,
  projectName: string,
) => {
  const project = readProjectConfigurationUnqualified(tree, projectName);

  const tsConfigPath = joinPathFragments(project.root, 'tsconfig.json');
  if (!tree.exists(tsConfigPath)) {
    throw new Error(
      `Selected plugin project ${projectName} is not a TypeScript project`,
    );
  }

  // Nx Plugins must use commonjs as a limitation of nx
  // https://github.com/nrwl/nx/issues/15682
  updateJson(tree, tsConfigPath, (tsConfig) => {
    tsConfig.compilerOptions ??= {};
    tsConfig.compilerOptions.module ??= 'commonjs';
    return tsConfig;
  });

  // Create an empty generators.json if one dosn't exist
  const generatorsJsonPath = joinPathFragments(project.root, 'generators.json');
  if (!tree.exists(generatorsJsonPath)) {
    writeJson(tree, generatorsJsonPath, {
      generators: {},
    });
  }

  // Create a package.json if one doesn't exist, and configure "main" and "generators"
  const pluginPackageJsonPath = joinPathFragments(project.root, 'package.json');
  if (!tree.exists(pluginPackageJsonPath)) {
    writeJson(tree, pluginPackageJsonPath, {
      name: project.name,
    });
  }
  updateJson(tree, pluginPackageJsonPath, (pkg) => {
    pkg.main ??= './src/index.js';
    pkg.generators ??= './generators.json';
    return pkg;
  });

  const rootPackageJson = tree.exists('package.json')
    ? readJson(tree, 'package.json')
    : undefined;
  const isNxPluginForAws = rootPackageJson?.name === '@aws/nx-plugin-source';

  if (!isNxPluginForAws) {
    // Add the required dependencies to the root package.json, and project's package.json
    const deps = {
      ...withVersions(['@nx/devkit']),
      // Include dependency on @aws/nx-plugin
      [PackageJson.name]: `^${PackageJson.version}`,
    };
    addDependenciesToPackageJson(tree, {}, deps);
    addDependenciesToPackageJson(tree, deps, {}, pluginPackageJsonPath);
  }
};
