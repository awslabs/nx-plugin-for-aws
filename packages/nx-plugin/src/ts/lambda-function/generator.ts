/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  addDependenciesToPackageJson,
  generateFiles,
  GeneratorCallback,
  installPackagesTask,
  joinPathFragments,
  OverwriteStrategy,
  ProjectConfiguration,
  Tree,
  updateJson,
  updateProjectConfiguration,
} from '@nx/devkit';
import { TsLambdaFunctionGeneratorSchema } from './schema';
import { sharedConstructsGenerator } from '../../utils/shared-constructs';
import {
  PACKAGES_DIR,
  SHARED_CONSTRUCTS_DIR,
} from '../../utils/shared-constructs-constants';
import { toClassName, toKebabCase, pascalCase } from '../../utils/names';
import { addStarExport } from '../../utils/ast';
import { formatFilesInSubtree } from '../../utils/format';
import { sortObjectKeys } from '../../utils/object';
import {
  NxGeneratorInfo,
  getGeneratorInfo,
  readProjectConfigurationUnqualified,
} from '../../utils/nx';
import { addGeneratorMetricsIfApplicable } from '../../utils/metrics';
import { withVersions } from '../../utils/versions';
import camelCase from 'lodash.camelcase';
import { TS_HANDLER_RETURN_TYPES } from './io';
import { addEsbuildBundleTarget } from '../../utils/esbuild';

export const TS_LAMBDA_FUNCTION_GENERATOR_INFO: NxGeneratorInfo =
  getGeneratorInfo(__filename);

/**
 * Generates a TypeScript Lambda Function to add to a TypeScript project
 */
export const tsLambdaFunctionGenerator = async (
  tree: Tree,
  schema: TsLambdaFunctionGeneratorSchema,
): Promise<GeneratorCallback> => {
  const projectConfig = readProjectConfigurationUnqualified(
    tree,
    schema.project,
  );

  const tsconfigPath = joinPathFragments(projectConfig.root, 'tsconfig.json');

  // Check if the project has a tsconfig.json file
  if (!tree.exists(tsconfigPath)) {
    throw new Error(
      `This generator does not support selected project ${schema.project}. The project must be a typescript project (ie contain a tsconfig.json)`,
    );
  }

  if (!projectConfig.sourceRoot) {
    throw new Error(
      `This project does not have a source root. Please add a source root to the project configuration before running this generator.`,
    );
  }

  const dir = projectConfig.root;
  const projectNameWithOutScope = projectConfig.name.split('/').pop();
  const projectNameKebabCase = toKebabCase(projectNameWithOutScope);
  const functionNameKebabCase = toKebabCase(schema.functionName);

  const constructFunctionNameKebabCase = `${projectNameKebabCase}-${functionNameKebabCase}`;
  const constructFunctionClassName = toClassName(
    constructFunctionNameKebabCase,
  );
  const lambdaFunctionCamelCase = camelCase(schema.functionName);
  const lambdaFunctionClassName = pascalCase(schema.functionName);
  const lambdaFunctionKebabCase = toKebabCase(schema.functionName);

  const functionPath = joinPathFragments(
    projectConfig.sourceRoot,
    schema.functionPath ?? '',
    `${lambdaFunctionKebabCase}.ts`,
  );

  // Check that the project does not already have a lambda handler
  if (tree.exists(functionPath)) {
    throw new Error(
      `This project already has a lambda function with the name ${functionNameKebabCase}. Please remove the lambda function before running this generator or use a different name.`,
    );
  }

  await sharedConstructsGenerator(tree);

  // Add bundle-<name> target for this specific lambda function
  const bundleTargetName = `bundle-${lambdaFunctionKebabCase}`;

  const enhancedOptions = {
    ...schema,
    dir,
    constructFunctionClassName,
    constructFunctionNameKebabCase,
    lambdaFunctionCamelCase,
    lambdaFunctionClassName,
    lambdaFunctionKebabCase,
    bundleTargetName,
    returnType: TS_HANDLER_RETURN_TYPES[schema.eventSource],
  };

  // Add a bundle target for the function
  addEsbuildBundleTarget(projectConfig, {
    bundleTargetName,
    targetFilePath: functionPath,
    extraEsbuildArgs: '--external:@aws-sdk/*',
  });

  projectConfig.targets = sortObjectKeys(projectConfig.targets);
  updateProjectConfiguration(tree, projectConfig.name, projectConfig);

  // Generate the lambda handler file
  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files', 'handler'),
    joinPathFragments(projectConfig.sourceRoot, schema.functionPath ?? ''),
    enhancedOptions,
    { overwriteStrategy: OverwriteStrategy.KeepExisting },
  );

  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files', SHARED_CONSTRUCTS_DIR, 'src', 'app'),
    joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR, 'src', 'app'),
    enhancedOptions,
    { overwriteStrategy: OverwriteStrategy.KeepExisting },
  );

  addStarExport(
    tree,
    joinPathFragments(
      PACKAGES_DIR,
      SHARED_CONSTRUCTS_DIR,
      'src',
      'app',
      'index.ts',
    ),
    './lambda-functions/index.js',
  );
  addStarExport(
    tree,
    joinPathFragments(
      PACKAGES_DIR,
      SHARED_CONSTRUCTS_DIR,
      'src',
      'app',
      'lambda-functions',
      'index.ts',
    ),
    `./${constructFunctionNameKebabCase}.js`,
  );

  // Ensure common constructs builds after our lambda function project
  updateJson(
    tree,
    joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR, 'project.json'),
    (config: ProjectConfiguration) => {
      if (!config.targets) {
        config.targets = {};
      }
      if (!config.targets.build) {
        config.targets.build = {};
      }
      config.targets.build.dependsOn = [
        ...(config.targets.build.dependsOn ?? []).filter(
          (t) => t !== `${projectConfig.name}:build`,
        ),
        `${projectConfig.name}:build`,
      ];
      return config;
    },
  );

  addDependenciesToPackageJson(
    tree,
    withVersions([
      '@aws-lambda-powertools/tracer',
      '@aws-lambda-powertools/logger',
      '@aws-lambda-powertools/metrics',
      '@aws-lambda-powertools/parser',
      '@middy/core',
      'zod',
    ]),
    withVersions(['esbuild', '@types/aws-lambda']),
  );

  await addGeneratorMetricsIfApplicable(tree, [
    TS_LAMBDA_FUNCTION_GENERATOR_INFO,
  ]);

  await formatFilesInSubtree(tree);

  return () => {
    installPackagesTask(tree);
  };
};

export default tsLambdaFunctionGenerator;
