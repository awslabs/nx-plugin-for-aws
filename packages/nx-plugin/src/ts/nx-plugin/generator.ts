/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  GeneratorCallback,
  OverwriteStrategy,
  Tree,
  generateFiles,
  installPackagesTask,
  joinPathFragments,
  updateJson,
  updateProjectConfiguration,
} from '@nx/devkit';
import { TsNxPluginGeneratorSchema } from './schema';
import {
  NxGeneratorInfo,
  getGeneratorInfo,
  readProjectConfigurationUnqualified,
} from '../../utils/nx';
import { addGeneratorMetricsIfApplicable } from '../../utils/metrics';
import { formatFilesInSubtree } from '../../utils/format';
import tsProjectGenerator, { getTsLibDetails } from '../lib/generator';
import { configureTsProjectAsNxPlugin } from './utils';
import tsMcpServerGenerator from '../mcp-server/generator';

export const TS_NX_PLUGIN_GENERATOR_INFO: NxGeneratorInfo =
  getGeneratorInfo(__filename);

export const tsNxPluginGenerator = async (
  tree: Tree,
  options: TsNxPluginGeneratorSchema,
): Promise<GeneratorCallback> => {
  // Generate a TypeScript project as the base
  const { fullyQualifiedName, dir } = getTsLibDetails(tree, options);
  await tsProjectGenerator(tree, options);

  // Configure the typescript project as an Nx Plugin
  configureTsProjectAsNxPlugin(tree, fullyQualifiedName);

  // Add a "package" target which creates a package which can be published to NPM
  const project = readProjectConfigurationUnqualified(tree, fullyQualifiedName);
  project.targets = {
    ...project.targets,
    package: {
      executor: '@nx/js:tsc',
      outputs: ['{options.outputPath}'],
      options: {
        outputPath: `dist/${dir}/package`,
        main: `${dir}/src/index.ts`,
        tsConfig: `${dir}/tsconfig.lib.json`,
        assets: [
          `${dir}/*.md`,
          `${dir}/LICENSE*`,
          `${dir}/NOTICE`,
          {
            input: `./${dir}/src`,
            glob: '**/!(*.ts)',
            output: './src',
          },
          {
            input: `./${dir}/src`,
            glob: '**/*.d.ts',
            output: './src',
          },
          {
            input: `./${dir}`,
            glob: 'generators.json',
            output: '.',
          },
          {
            input: `./${dir}`,
            glob: 'executors.json',
            output: '.',
          },
        ],
      },
    },
    build: {
      dependsOn: [
        ...(project.targets?.build?.dependsOn ?? []).filter(
          (d) => typeof d !== 'string' || d !== 'package',
        ),
        'package',
      ],
    },
  };
  updateProjectConfiguration(tree, fullyQualifiedName, project);

  // Remove the hello world example from index.ts
  tree.write(joinPathFragments(project.sourceRoot, 'index.ts'), '');

  // Add an MCP Server
  await tsMcpServerGenerator(tree, { project: fullyQualifiedName });

  const mcpPath = joinPathFragments(project.sourceRoot, 'mcp-server');
  const mcpServerPath = joinPathFragments(mcpPath, 'server.ts');

  // Remove the sample server and tools
  tree.delete(mcpServerPath);
  tree.delete(joinPathFragments(mcpPath, 'tools', 'add.ts'));
  tree.delete(joinPathFragments(mcpPath, 'resources', 'sample-guidance.ts'));

  // Add generator MCP server
  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files', 'mcp-server'),
    mcpPath,
    {
      name: fullyQualifiedName,
    },
    { overwriteStrategy: OverwriteStrategy.KeepExisting },
  );

  await addGeneratorMetricsIfApplicable(tree, [TS_NX_PLUGIN_GENERATOR_INFO]);

  await formatFilesInSubtree(tree);
  return () => {
    installPackagesTask(tree);
  };
};

export default tsNxPluginGenerator;
