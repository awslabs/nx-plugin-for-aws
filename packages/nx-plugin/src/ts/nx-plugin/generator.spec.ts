/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Tree, readProjectConfiguration, readJson } from '@nx/devkit';
import { tsNxPluginGenerator, TS_NX_PLUGIN_GENERATOR_INFO } from './generator';
import {
  createTreeUsingTsSolutionSetup,
  snapshotTreeDir,
} from '../../utils/test';
import { expectHasMetricTags } from '../../utils/metrics.spec';
import { sharedConstructsGenerator } from '../../utils/shared-constructs';

describe('ts#nx-plugin generator', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTreeUsingTsSolutionSetup();
  });

  it('should configure the project as an Nx Plugin with correct targets', async () => {
    await tsNxPluginGenerator(tree, { name: 'test-plugin' });

    const project = readProjectConfiguration(tree, '@proj/test-plugin');

    // Should have package target
    expect(project.targets?.package).toBeDefined();
    expect(project.targets?.package?.executor).toBe('@nx/js:tsc');
    expect(project.targets?.package?.outputs).toEqual(['{options.outputPath}']);
    expect(project.targets?.package?.options?.outputPath).toBe(
      'dist/test-plugin/package',
    );
    expect(project.targets?.package?.options?.main).toBe(
      'test-plugin/src/index.ts',
    );
    expect(project.targets?.package?.options?.tsConfig).toBe(
      'test-plugin/tsconfig.lib.json',
    );

    // Should have correct assets
    const assets = project.targets?.package?.options?.assets;
    expect(assets).toContain('test-plugin/*.md');
    expect(assets).toContain('test-plugin/LICENSE*');
    expect(assets).toContain('test-plugin/NOTICE');
    expect(assets).toContainEqual({
      input: './test-plugin/src',
      glob: '**/!(*.ts)',
      output: './src',
    });
    expect(assets).toContainEqual({
      input: './test-plugin/src',
      glob: '**/*.d.ts',
      output: './src',
    });
    expect(assets).toContainEqual({
      input: './test-plugin',
      glob: 'generators.json',
      output: '.',
    });
    expect(assets).toContainEqual({
      input: './test-plugin',
      glob: 'executors.json',
      output: '.',
    });
  });

  it('should configure build target to depend on package', async () => {
    await tsNxPluginGenerator(tree, { name: 'test-plugin' });

    const project = readProjectConfiguration(tree, '@proj/test-plugin');

    expect(project.targets?.build?.dependsOn).toContain('package');
  });

  it('should clear the index.ts file', async () => {
    await tsNxPluginGenerator(tree, { name: 'test-plugin' });

    const indexContent = tree.read('test-plugin/src/index.ts')?.toString();
    expect(indexContent).toBe('');
  });

  it('should configure TypeScript project as Nx Plugin', async () => {
    await tsNxPluginGenerator(tree, { name: 'test-plugin' });

    // Check tsconfig.json has commonjs module
    const tsConfig = readJson(tree, 'test-plugin/tsconfig.json');
    expect(tsConfig.compilerOptions?.module).toBe('commonjs');

    // Check generators.json exists
    expect(tree.exists('test-plugin/generators.json')).toBe(true);
    const generatorsJson = readJson(tree, 'test-plugin/generators.json');
    expect(generatorsJson.generators).toBeDefined();

    // Check package.json configuration
    expect(tree.exists('test-plugin/package.json')).toBe(true);
    const packageJson = readJson(tree, 'test-plugin/package.json');
    expect(packageJson.main).toBe('./src/index.js');
    expect(packageJson.generators).toBe('./generators.json');
  });

  it('should generate MCP server files', async () => {
    await tsNxPluginGenerator(tree, { name: 'test-plugin' });

    const project = readProjectConfiguration(tree, '@proj/test-plugin');
    const mcpPath = `${project.sourceRoot}/mcp-server`;

    // Should generate MCP server files
    expect(tree.exists(`${mcpPath}/server.ts`)).toBe(true);
    expect(tree.exists(`${mcpPath}/tools/general-guidance.ts`)).toBe(true);
    expect(tree.exists(`${mcpPath}/tools/create-workspace-command.ts`)).toBe(
      true,
    );
    expect(tree.exists(`${mcpPath}/tools/generator-guide.ts`)).toBe(true);

    snapshotTreeDir(tree, mcpPath);
  });

  it('should remove sample MCP server files', async () => {
    await tsNxPluginGenerator(tree, { name: 'test-plugin' });

    const project = readProjectConfiguration(tree, '@proj/test-plugin');
    const mcpPath = `${project.sourceRoot}/mcp-server`;

    // Should not have sample files
    expect(tree.exists(`${mcpPath}/tools/add.ts`)).toBe(false);
    expect(tree.exists(`${mcpPath}/resources/sample-guidance.ts`)).toBe(false);
  });

  it('should add required dependencies', async () => {
    // Modify package.json to simulate external workspace
    tree.write(
      'package.json',
      JSON.stringify({
        name: 'external-workspace',
        version: '1.0.0',
      }),
    );

    await tsNxPluginGenerator(tree, { name: 'test-plugin' });

    const rootPackageJson = readJson(tree, 'package.json');
    expect(rootPackageJson.devDependencies?.['@nx/devkit']).toBeDefined();
    expect(rootPackageJson.devDependencies?.['@aws/nx-plugin']).toBeDefined();

    const projectPackageJson = readJson(tree, 'test-plugin/package.json');
    expect(projectPackageJson.dependencies?.['@nx/devkit']).toBeDefined();
    expect(projectPackageJson.dependencies?.['@aws/nx-plugin']).toBeDefined();
  });

  it('should add generator metric to app.ts', async () => {
    await sharedConstructsGenerator(tree);

    await tsNxPluginGenerator(tree, { name: 'my-plugin' });

    expectHasMetricTags(tree, TS_NX_PLUGIN_GENERATOR_INFO.metric);
  });
});
