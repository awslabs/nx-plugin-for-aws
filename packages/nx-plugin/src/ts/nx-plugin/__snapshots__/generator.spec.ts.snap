// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ts#nx-plugin generator > should generate MCP server files > test-plugin/src/mcp-server/http.ts 1`] = `
"#!/usr/bin/env node
import { createServer } from './server';
import { StreamableHTTPServerTransport } from '@modelcontextprotocol/sdk/server/streamableHttp.js';
import express, { Request, Response } from 'express';

const PORT = 8000;

const app = express();
app.use(express.json());

app.post('/mcp', async (req: Request, res: Response) => {
  try {
    const server = createServer();
    const transport: StreamableHTTPServerTransport =
      new StreamableHTTPServerTransport({
        sessionIdGenerator: undefined,
      });
    res.on('close', () => {
      console.log('Request closed');
      transport.close();
      server.close();
    });
    await server.connect(transport);
    await transport.handleRequest(req, res, req.body);
  } catch (error) {
    console.error('Error handling MCP request:', error);
    if (!res.headersSent) {
      res.status(500).json({
        jsonrpc: '2.0',
        error: {
          code: -32603,
          message: 'Internal server error',
        },
        id: null,
      });
    }
  }
});

// SSE notifications not supported in stateless mode
app.get('/mcp', async (req: Request, res: Response) => {
  console.log('Received GET MCP request');
  res.writeHead(405).end(
    JSON.stringify({
      jsonrpc: '2.0',
      error: {
        code: -32000,
        message: 'Method not allowed.',
      },
      id: null,
    }),
  );
});

// Session termination not needed in stateless mode
app.delete('/mcp', async (req: Request, res: Response) => {
  console.log('Received DELETE MCP request');
  res.writeHead(405).end(
    JSON.stringify({
      jsonrpc: '2.0',
      error: {
        code: -32000,
        message: 'Method not allowed.',
      },
      id: null,
    }),
  );
});

// Start the server
void (async () => {
  app.listen(PORT, (error) => {
    if (error) {
      console.error('Failed to start server:', error);
      process.exit(1);
    }
    console.log(
      \`MCP Stateless Streamable HTTP Server listening on port \${PORT}\`,
    );
  });
})();
"
`;

exports[`ts#nx-plugin generator > should generate MCP server files > test-plugin/src/mcp-server/index.ts 1`] = `
"export * from './server';
"
`;

exports[`ts#nx-plugin generator > should generate MCP server files > test-plugin/src/mcp-server/resources/GENERAL_GUIDANCE.md 1`] = `
"# General Guidance

## Tool Selection Guide

- Use the \`general-guidance\` tool for guidance and best practices for working with Nx and @proj/test-plugin.
- Use the \`create-workspace-command\` tool to discover how to create a workspace to start a new project.
- Use the \`list-generators\` tool to discover the available generators and how to run them.
- Use the \`generator-guide\` tool to retrieve detailed information about a specific generator.

## Getting Started

- Choose a package manager first. You can choose between pnpm, bun, yarn or npm. Check if one is already in use if working in an existing project. It's recommended to use "pnpm" if the user has no preference
- If starting from scratch, you must create an Nx workspace. Use the \`create-workspace-command\` tool for more details, and provide it with your chosen package manager
- After this, you can start scaffolding the main components of your application using generators. Use the \`list-generators\` tool to discover available generators, and the \`generator-guide\` tool for more detailed information about a specific generator

## Nx Primer

- Prefix nx commands with the appropriate prefix for your package manager, for example:
  - pnpm nx <options>
  - bunx nx <options>
  - npx nx <options>
  - yarn nx <options>
- Each project in your workspace has a file named \`project.json\` which contains important project information such as its name, and defines the "targets" which can be run for that project, for example building or testing the project
- Use the command \`nx reset\` to reset the Nx daemon when unexpected issues arise
- After adding dependencies between TypeScript projects, use \`nx sync\` to ensure project references are set up correctly

## General Instructions

- Workspaces contain a single \`package.json\` file at the root which defines the dependencies for all projects. Therefore when installing dependencies, you must add these to the root workspace using the appropriate command for your package manager:
  - pnpm add -w -D <package>
  - yarn add -D <package>
  - npm install --legacy-peer-deps -D <package>
  - bun install -D <package>
  - (Omit -D for production dependencies)
- When specifying project names as arguments to generators, prefer the _fully qualified_ project name, for example \`@workspace-name/project-name\`. Check the \`project.json\` file for the specific package to find its fully qualified name
- When no generator exists for a specific framework required, look for more generic project generators if availalbe and build on top of these.

## Useful Commands

- Fix lint issues with \`nx run-many --target lint --configuration=fix --all --output-style=stream\`
- Build all projects with \`nx run-many --target build --all --output-style=stream\`
- Prefer importing the CDK constructs vended by generators in \`packages/common/constructs\` over writing your own

## Best Practices

- After running a generator, use the \`nx show projects\` command to check which projects have been added (if any)
- Carefully examine the files that have been generated and always refer back to the generator guide when working in a generated project
- Generate all projects into the \`packages/\` directory
- After making changes to your projects, fix linting issues, then run a full build
"
`;

exports[`ts#nx-plugin generator > should generate MCP server files > test-plugin/src/mcp-server/schema.ts 1`] = `
"import { z } from 'zod-v3';

export const PACKAGE_MANAGERS = ['pnpm', 'yarn', 'npm', 'bun'] as const;
export const PackageManagerSchema = z.enum(PACKAGE_MANAGERS);
"
`;

exports[`ts#nx-plugin generator > should generate MCP server files > test-plugin/src/mcp-server/server.ts 1`] = `
"import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { registerCreateWorkspaceCommandTool } from './tools/create-workspace-command';
import { registerGeneralGuidanceTool } from './tools/general-guidance';
import { registerListGeneratorsTool } from './tools/list-generators';
import { registerGeneratorGuideTool } from './tools/generator-guide';
import { getPackageName } from './utils';

/**
 * Create the MCP Server
 */
export const createServer = () => {
  const packageName = getPackageName();

  const server = new McpServer(
    {
      name: packageName,
      version: '1.0.0',
    },
    {
      instructions: \`# \${packageName} MCP Server

This server provides resources and tools for quickly scaffolding projects using \${packageName}.

\${packageName} provides "generators" to add projects or functionality to your workspace. Use this to build the foundations of any project you are building, if the generators apply to your use case.

## Tool Selection Guide

- Use the \\\`general-guidance\\\` tool for guidance and best practices for working with Nx and \${packageName}.
- Use the \\\`create-workspace-command\\\` tool to discover how to create a workspace to start a new project.
- Use the \\\`list-generators\\\` tool to discover the available generators and how to run them.
- Use the \\\`generator-guide\\\` tool to retrieve detailed information about a specific generator.\`,
    },
  );

  registerCreateWorkspaceCommandTool(server);
  registerGeneralGuidanceTool(server);
  registerListGeneratorsTool(server);
  registerGeneratorGuideTool(server);

  return server;
};
"
`;

exports[`ts#nx-plugin generator > should generate MCP server files > test-plugin/src/mcp-server/stdio.ts 1`] = `
"#!/usr/bin/env node
import { StdioServerTransport } from '@modelcontextprotocol/sdk/server/stdio.js';
import { createServer } from './server';

export const startMcpServer = async () => {
  const transport = new StdioServerTransport();
  await createServer().connect(transport);
  console.error('MCP Server listening on STDIO');
};

void (async () => {
  try {
    await startMcpServer();
  } catch (e) {
    console.error(e);
  }
})();
"
`;

exports[`ts#nx-plugin generator > should generate MCP server files > test-plugin/src/mcp-server/tools/create-workspace-command.ts 1`] = `
"import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { z } from 'zod-v3';
import { PackageManagerSchema } from '../schema';
import { buildInstallCommand, getPackageName } from '../utils';

/**
 * Registers a tool which tells a model how to create an Nx workspace
 */
export const registerCreateWorkspaceCommandTool = (server: McpServer) => {
  server.tool(
    'create-workspace-command',
    'Tool to discover how to create a workspace to start a new project.',
    { workspaceName: z.string(), packageManager: PackageManagerSchema },
    ({ workspaceName, packageManager }) => ({
      content: [
        {
          type: 'text',
          text: \`Run the following command to create a workspace:

\\\`\\\`\\\`bash
npx create-nx-workspace@~21.0.3 \${workspaceName} --pm=\${packageManager} --preset=@aws/nx-plugin --ci=skip
\\\`\\\`\\\`

This will create a new workspace within the \${workspaceName} directory.

Then, install the plugin package to ensure generators are available:

\\\`\\\`\\\`bash
\${buildInstallCommand(packageManager, getPackageName())}
\\\`\\\`\\\`.
  \`,
        },
      ],
    }),
  );
};
"
`;

exports[`ts#nx-plugin generator > should generate MCP server files > test-plugin/src/mcp-server/tools/general-guidance.ts 1`] = `
"import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import * as fs from 'fs';
import * as path from 'path';

/**
 * Registers a tool which provides general guidance for using Nx and this plugin
 */
export const registerGeneralGuidanceTool = (server: McpServer) => {
  server.tool(
    'general-guidance',
    'Tool for guidance and best practices for working with Nx and this plugin',
    async () => ({
      content: [
        {
          type: 'text',
          text: fs.readFileSync(
            path.resolve(__dirname, '../resources/GENERAL_GUIDANCE.md'),
            'utf-8',
          ),
        },
      ],
    }),
  );
};
"
`;

exports[`ts#nx-plugin generator > should generate MCP server files > test-plugin/src/mcp-server/tools/generator-guide.ts 1`] = `
"import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { z } from 'zod-v3';
import { renderGeneratorInfo, listGenerators } from '../utils';
import { PackageManagerSchema } from '../schema';
import * as fs from 'fs';
import * as path from 'path';

/**
 * Registers a tool which provides a detailed guide for an individual generator
 */
export const registerGeneratorGuideTool = (server: McpServer) => {
  server.tool(
    'generator-guide',
    'Tool to retrieve detailed information about a specific generator.',
    {
      packageManager: PackageManagerSchema,
      generator: z.string(),
    },
    async ({ packageManager, generator: generatorId }) => {
      const generators = listGenerators();
      const generator = generators.find((g) => g.id === generatorId);
      if (!generator) {
        throw new Error(
          \`No generator found with id \${generatorId}. Available generators: \${generators.map((g) => g.id).join(' ,')}\`,
        );
      }

      return {
        content: [
          {
            type: 'text',
            text: \`## \${renderGeneratorInfo(packageManager, generator)}

# Guide

\${fs.readFileSync(path.join(generator.resolvedPath, 'README.md'))}
\`,
          },
        ],
      };
    },
  );
};
"
`;

exports[`ts#nx-plugin generator > should generate MCP server files > test-plugin/src/mcp-server/tools/list-generators.ts 1`] = `
"import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { PackageManagerSchema } from '../schema';
import { renderGeneratorInfo, listGenerators } from '../utils';

/**
 * Registers a tool which lists details about the available generators
 */
export const registerListGeneratorsTool = (server: McpServer) => {
  server.tool(
    'list-generators',
    'Tool to discover the available generators and how to run them.',
    { packageManager: PackageManagerSchema },
    ({ packageManager }) => ({
      content: [
        {
          type: 'text',
          text: \`## Available Generators

  \${listGenerators()
    .map((g) => \`### \${renderGeneratorInfo(packageManager, g)}\`)
    .join('\\n\\n')}
  \`,
        },
      ],
    }),
  );
};
"
`;

exports[`ts#nx-plugin generator > should generate MCP server files > test-plugin/src/mcp-server/utils.ts 1`] = `
"import * as fs from 'fs';
import * as path from 'path';

export interface NxGeneratorInfo {
  readonly id: string;
  readonly resolvedPath: string;
  readonly description?: string;
}

export const getPackageName = (): string =>
  JSON.parse(
    fs.readFileSync(path.resolve(__dirname, '../../package.json'), 'utf-8'),
  ).name;

export const listGenerators = (): NxGeneratorInfo[] =>
  Object.entries(
    JSON.parse(
      fs.readFileSync(
        path.resolve(__dirname, '../../generators.json'),
        'utf-8',
      ),
    ).generators,
  ).map(([id, info]: any) => ({
    id,
    resolvedPath: path.resolve(__dirname, '..', '..', info.factory, '..'),
    description: info.description,
  }));

/**
 * Build a command to run nx
 */
export const buildNxCommand = (command: string, pm?: string) =>
  \`\${
    pm
      ? \`\${
          {
            npm: 'npx',
            bun: 'bunx',
          }[pm] ?? pm
        } \`
      : ''
  }nx \${command}\`;

export const buildInstallCommand = (pm: string, pkg: string) =>
  ({
    pnpm: \`pnpm add -Dw \${pkg}\`,
    yarn: \`yarn add -D \${pkg}\`,
    npm: \`npm install --legacy-peer-deps -D \${pkg}\`,
    bun: \`bun install -D \${pkg}\`,
  })[pm];

const renderSchema = (schema: any) =>
  Object.entries(schema.properties)
    .map(
      ([parameter, parameterSchema]: [string, any]) =>
        \`- \${parameter} [type: \${parameterSchema.type}]\${(schema.required ?? []).includes(parameter) ? \` (required)\` : ''} \${parameterSchema.description}\`,
    )
    .join('\\n');

const renderGeneratorCommand = (
  generatorId: string,
  schema: any,
  packageManager?: string,
) => \`\\\`\\\`\\\`bash
\${buildNxCommand(
  \`g \${getPackageName()}:\${generatorId} --no-interactive \${Object.entries(
    schema.properties,
  )
    .filter(([parameter]) => (schema.required ?? []).includes(parameter))
    .map(([parameter]: [string, any]) => \`--\${parameter}=<\${parameter}>\`)
    .join(' ')}\`,
  packageManager,
)}
\\\`\\\`\\\`\`;

/**
 * Render summary information about a generator
 */
export const renderGeneratorInfo = (
  packageManager: string,
  info: NxGeneratorInfo,
): string => {
  const schema = JSON.parse(
    fs.readFileSync(path.join(info.resolvedPath, 'schema.json'), 'utf-8'),
  );

  return \`\${info.id}

Description: \${info.description}

Available Parameters:
\${renderSchema(schema)}

Command:
\${renderGeneratorCommand(info.id, schema, packageManager)}
\`;
};
"
`;
